openrtb/bid_test.go

```diff
@@ -58,3 +61,86 @@ func TestBidValidation(t *testing.T) {
 		}
 	}
 }
+
+func TestBid_Copy(t *testing.T) {
+	testCases := []struct {
+		bid *openrtb.Bid
+	}{
+		{
+			&openrtb.Bid{},
+		},
+		{
+			&openrtb.Bid{
+				ID:                 "test",
+				ImpressionID:       "testImpId",
+				Price:              0,
+				AdID:               "testAdID",
+				WinNotificationURL: "testWinNURL",
+				AdMarkup:           "testAdm",
+				AdvertiserDomains:  []string{},
+				Bundle:             "testBundle",
+				QualityImageURL:    "testQualityImgURL",
+				CampaignID:         "testCID",
+				CreativeID:         "testCrID",
+				Categories:         []openrtb.Category{},
+				CreativeAttributes: []openrtb.CreativeAttribute{},
+				DealID:             "testDealID",
+				Height:             0,
+				Width:              0,
+			},
+		},
+		{
+			&openrtb.Bid{
+				ID:                 "test",
+				ImpressionID:       "testImpId",
+				Price:              0,
+				AdID:               "testAdID",
+				WinNotificationURL: "testWinNURL",
+				AdMarkup:           "testAdm",
+				AdvertiserDomains:  []string{"123"},
+				Bundle:             "testBundle",
+				QualityImageURL:    "testQualityImgURL",
+				CampaignID:         "testCID",
+				CreativeID:         "testCrID",
+				Categories:         []openrtb.Category{openrtb.CategoryArtsEntertainment},
+				CreativeAttributes: []openrtb.CreativeAttribute{openrtb.CreativeAttributeAudioAuto},
+				DealID:             "testDealID",
+				Height:             0,
+				Width:              0,
+				Extension:          json.RawMessage([]byte(`rawr`)),
+			},
+		},
+	}
+	for _, testCase := range testCases {
+		b2 := testCase.bid.Copy()
+
+		if b2 == testCase.bid {
+			t.Errorf("Address of bid should not be the same. b1: %v b2: %v.", &testCase.bid, &b2)
+		}
+
+		if !reflect.DeepEqual(testCase.bid, b2) {
+			b1JSON, _ := json.MarshalIndent(testCase.bid, "", "  ")
+			b2JSON, _ := json.MarshalIndent(b2, "", "  ")
+			t.Errorf("Bids should hold the same values.\nExpected: %s\n Got: %s", b1JSON, b2JSON)
+		}
+
+		b2.ID = "1234"
+		if testCase.bid.ID == "1234" {
+			t.Errorf("Bid b2 should not be pointing to original bid object.\ntestCase: %+v\nCopy: %+v", testCase.bid, b2)
+		}
+
+		if len(b2.AdvertiserDomains) > 0 {
+			b2.AdvertiserDomains[0] = "456"
+			if testCase.bid.AdvertiserDomains[0] == "456" {
```

Comment:
i like the address checking from below better, but up to you if you want to change it, no-op is fine

Reply:
can I check the address? it's not a slice of pointers.